### PR TITLE
API to merge offline regions from an external database

### PR DESCRIPTION
**Pull request by asheemmamoowala**
Addresses #3894.
Possibly addresses #9947.

This PR introduces a new core API to merge offline regions from an externally prepared database into the main offline database for an application. The secondary database is best prepared using the `mbgl-filesource` module, for example, using the `mbgl-offline` tool.

`DefaultFileSource::mergeOfflineRegions` will merge all regions and their corresponding tiles and resources from an external database into the main offline database. Duplicate regions are flattened into a single new region, and resources that already exist are updated if newer (`modified` more recently).

### TODO:
- [x] Migrate old versions of the external database if needed 
- [x] Allow multiple concurrent calls (is this needed)
- [x] Tests
- [x] Investigate adding INDEXes for the temp tables used for merging.
- [x] Documentation
- [ ] Update [sideloading wiki](https://github.com/mapbox/mapbox-gl-native/wiki/Sideloading-offline-maps)

cc @kkaefer @ericdeveloper

### CODE REVIEW

**Comment by 1ec5** [comment]
> Duplicate regions are flattened into a single new region

If each region has metadata associated with it, which region’s metadata is preserved? Should there be a conflict resolution mechanism?

**Comment by asheemmamoowala** [comment]
@1ec5 Regions are duplicate only if the definition and description (aka metadata) are equal.

**Comment by 1ec5** [comment]
> Regions are duplicate only if the definition and description (aka metadata) are equal.

OK, this sounds fine as far as the iOS and macOS map SDKs are concerned. MGLOfflineStorage accepts the metadata in the form of an NSData object, but what it passes to mbgl as the metadata is the raw bytes contained by the NSData object, not the object itself. That means a comparison between two `mbgl::OfflineRegionMetadata` instances would be equivalent to calling `-[NSData isEqualToData:]`. 👍

https://github.com/mapbox/mapbox-gl-native/blob/60cbedc344eff10df0069a9d243a555dc59992ee/platform/darwin/src/MGLOfflineStorage.mm#L289-L290

**Comment by jfirebaugh** [comment]
> Regions are duplicate only if the definition and description (aka metadata) are equal.

Where is this equality defined? In merge_sideloaded.sql it looks like only the `definition` columns are compared.

Is de-duplicating regions a requirement, or could we always just retain all sideloaded regions? Doing it via bytewise comparison seems fraught, because regions that are logically identical may not always be bytewise equal. I can't think of a really strong example given the current code, but if we update RapidJSON, change the JSON key ordering, add fields, or otherwise change the serialization format, all of those might lead to issues in the future.

**Comment by asheemmamoowala** [comment]
> Where is this equality defined? In merge_sideloaded.sql it looks like only the definition columns are compared.

Good catch @jfirebaugh - I noticed that it wasn't comparing against descriptions until after @1ec5 brought it up, and it is incoming. 

The de-duplication is neither a requirement nor a _feature_, it just needs to happen because there are no unique fields in the `regions` table. Without de-duplication, I cant think of a way to map regions from the sideloaded db to their inserted counterparts in the main db.

See: https://gist.github.com/asheemmamoowala/457170411eb98dab68375bfc9d90743e

File: bin/offline.cpp [hunk 1]
```diff
@@ -61,19 +61,20 @@ int main(int argc, char *argv[]) {
     args::ValueFlag<std::string> styleValue(argumentParser, "URL", "Map stylesheet", {'s', "style"});
     args::ValueFlag<std::string> outputValue(argumentParser, "file", "Output database file name", {'o', "output"});
     args::ValueFlag<std::string> apiBaseValue(argumentParser, "URL", "API Base URL", {'a', "apiBaseURL"});
-    
+    args::ValueFlag<std::string> mergePathValue(argumentParser, "merge", "Databse to merge from", {'m', "merge"});
```
Comment (kkaefer) [inline bin/offline.cpp]: typo

File: bin/offline.cpp [hunk 2]
```diff
@@ -61,19 +61,20 @@ int main(int argc, char *argv[]) {
...
+    args::ValueFlag<std::string> inputValue(argumentParser, "input", "Databse to merge into. Use with --merge option.", {'i', "input"});
```
Comment (kkaefer) [inline bin/offline.cpp]: typo

File: bin/offline.cpp [hunk 3]
```diff
@@ -136,15 +142,37 @@ int main(int argc, char *argv[]) {
     fileSource.setAccessToken(token);
     fileSource.setAPIBaseURL(apiBaseURL);
 
+    if (inputDb && mergePath) {
```
Comment (kkaefer) [inline bin/offline.cpp]: While this is not a high priority tool, can we add a small error message to let people know if they've only supplied one, but not both parameters?

File: platform/default/mbgl/storage/offline_database.cpp [hunk 4]
```diff
@@ -637,6 +638,82 @@ OfflineDatabase::createRegion(const OfflineRegionDefinition& definition,
...
+OfflineDatabase::mergeDatabase(const std::string& sideDatabasePath) {
+    try {
+        // clang-format off
+        mapbox::sqlite::Query query{ getStatement("ATTACH DATABASE ?1 AS side") };
+        // clang-format on
+
+        query.bind(1, sideDatabasePath);
+        query.run();
+    } catch (const mapbox::sqlite::Exception& ex) {
+        Log::Error(Event::Database, static_cast<int>(ex.code), "Can't attach database (%s) for merge: %s", sideDatabasePath.c_str(), ex.what());
```
Comment (kkaefer) [inline platform/default/mbgl/storage/offline_database.cpp]: Can we use `handleError` here instead? It has facilities for closing/reopening the database in some circumstances.

File: platform/default/mbgl/storage/offline_database.cpp [hunk 5]
```diff
@@ -637,6 +638,82 @@ OfflineDatabase::createRegion(const OfflineRegionDefinition& definition,
...
+                decodeOfflineRegionDefinition(queryRegions.get<std::string>(1)),
+                queryRegions.get<std::vector<uint8_t>>(2));
+            result.emplace_back(std::move(region));
+        }
+        db->exec("DETACH DATABASE side");
+        // Explicit move to avoid triggering the copy constructor.
+        return { std::move(result) };
+    } catch (const std::runtime_error& ex) {
+        db->exec("DETACH DATABASE side");
+        Log::Error(Event::Database, "%s", ex.what());
```
Comment (kkaefer) [inline platform/default/mbgl/storage/offline_database.cpp]: Same note about using `handleError` here as well.

File: platform/default/mbgl/storage/offline_database.cpp [hunk 6]
```diff
@@ -637,6 +638,82 @@ OfflineDatabase::createRegion(const OfflineRegionDefinition& definition,
...
+        Log::Error(Event::Database, static_cast<int>(ex.code), "Can't attach database (%s) for merge: %s", sideDatabasePath.c_str(), ex.what());
+        return unexpected<std::exception_ptr>(std::current_exception());
+    }
+    try {
+        // Support sideloaded databases at user_version = 6. Future schema version
+        // changes will need to implement migration paths for sideloaded databases at
+        // version 6.
+        auto sideUserVersion = static_cast<int>(getPragma<int64_t>("PRAGMA side.user_version"));
+        const auto mainUserVersion = getPragma<int64_t>("PRAGMA user_version");
+        if (sideUserVersion < 6 || sideUserVersion != mainUserVersion) {
```
Comment (kkaefer) [inline platform/default/mbgl/storage/offline_database.cpp]: Why do we check for `sideUserVersion < 6`?

File: platform/default/mbgl/storage/offline_database.cpp (hunk 5 above)
Comment (asheemmamoowala) [reply platform/default/mbgl/storage/offline_database.cpp]: If the main database issues are handled during the attach step ([comment](https://github.com/mapbox/mapbox-gl-native/pull/12608#discussion_r212352371)), then this try...catch block should not call `handleError`, which would remove the existing database. If there is an exception during the merge operation due to the sideloaded database having an invalid schema, this should rollback the transaction and attempt to detach the database.
Do you think there needs to be more checking for errors in the main database after a failed merge ?

File: platform/default/mbgl/storage/offline_database.cpp (hunk 4 above)
Comment (kkaefer) [reply platform/default/mbgl/storage/offline_database.cpp]: Ooh, good point.

File: platform/default/mbgl/storage/offline_database.cpp (hunk 6 above)
Comment (asheemmamoowala) [reply platform/default/mbgl/storage/offline_database.cpp]: The two `<6` condition is for the minimum schema version check. The `1+ mainUserVersion` can be switched to an else clause when we revise the schema version and support a forward migration of older side-loaded databases.

File: bin/offline.cpp (hunk 3 above)
Comment (asheemmamoowala) [reply bin/offline.cpp]: Addressed by using an `ArgumentsGroup`

**Comment by asheemmamoowala** [comment]
> We currently don't upgrade the sideloaded DB to the current version. Is that something that upstream users (aka platform bindings) have to handle before passing the filename to mergeDatabase?

The minimum sideload db version supported by this API is v6, so there is no need at this time for upgrading the database. I've added more documentation to indicate the responsibility of the SDK bindings to make sure the path is writable. Ideally implementations will copy the input path to a temporary folder and delete it after the operation is complete.

**Pull request merged**